src/System.Diagnostics.FileVersionInfo/src/System/Diagnostics/FileVersionInfo.Metadata.cs

```diff
@@ -167,6 +151,81 @@ private void LoadManagedAssemblyMetadata(MetadataReader metadataReader)
+            // both the assembly version and the product version in the Win32 resources. If it doesn't 
+            // see an [AssemblyVersion(...)], then it sets the assembly version to 0.0.0.0, however it 
+            // sets the product version in the Win32 resources to whatever was defined in the 
+            // [AssemblyFileVersionAttribute(...)] if there was one (unless there is an AssemblyInformationalVersionAttribute,
+            // in which case it always uses that for the product version).  Without parsing the Win32 resources,
+            // we can't differentiate these two cases, so given the rarity of explicitly setting an 
+            // assembly's version number to 0.0.0.0, we assume that if it is 0.0.0.0 then the attribute 
+            // wasn't specified and we use the file version.
+
+            if (!sawAssemblyInformationalVersionAttribute && _productVersion == "0.0.0.0")
+            {
+                _productVersion = _fileVersion;
+                _productMajor = _fileMajor;
+                _productMinor = _fileMinor;
+                _productBuild = _fileBuild;
+                _productPrivate = _filePrivate;
+            }
+        }
+
+        /// <summary>Parses the version into its constituent parts.</summary>
+        private static void ParseVersion(string versionString, out int major, out int minor, out int build, out int priv)
+        {
+            // Relatively-forgiving parsing of a version:
+            // - If there are more than four parts (separated by periods), all results are deemed 0
+            // - If any part fails to parse completely as an integer, no further parts are parsed and are left as 0.
+            // - If any part partially parses as an integer, that value is used for that part.
+            // - Whitespace is treated like any other non-digit character and thus isn't ignored.
+            // - Each component is parsed as a ushort, allowing for overflow.
+
+            string[] parts = versionString.Split(s_versionSeparators);
+            major = minor = build = priv = 0;
+            if (parts.Length <= 4)
+            {
+                bool endedEarly;
+                if (parts.Length > 0)
+                {
+                    major = ParseUInt16UntilNonDigit(parts[0], out endedEarly);
+                    if (!endedEarly && parts.Length > 1)
+                    {
+                        minor = ParseUInt16UntilNonDigit(parts[1], out endedEarly);
+                        if (!endedEarly && parts.Length > 2)
+                        {
+                            build = ParseUInt16UntilNonDigit(parts[2], out endedEarly);
+                            if (!endedEarly && parts.Length > 3)
+                            {
+                                priv = ParseUInt16UntilNonDigit(parts[3], out endedEarly);
+                            }
+                        }
+                    }
```

Comment:
Highly subjective but early returns might be easier to follow than this much nesting.